### PR TITLE
[JBIDE-21359] Add build tab to explorer deployment view

### PR DESCRIPTION
**Pull request by jcantrill**

### CODE REVIEW

File: plugins/org.jboss.tools.openshift.common.ui/src/org/jboss/tools/openshift/internal/common/ui/explorer/DefaultPropertySection.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+	@Override
+	public void setInput(IWorkbenchPart part, ISelection selection) {
+		super.setInput(part, selection);
+		page.selectionChanged(part, selection);
+	}
+
+	@Override
+	public void refresh() {
+		page.refresh();
```
Comment (fbricon) [inline plugins/org.jboss.tools.openshift.common.ui/src/org/jboss/tools/openshift/internal/common/ui/explorer/DefaultPropertySection.java]: shouldn't you check if page is disposed before refreshing?

File: plugins/org.jboss.tools.openshift.common.ui/src/org/jboss/tools/openshift/internal/common/ui/explorer/ExplorerViewTabbedPropertyAdapterFactory.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+/**
+ * An adapter factory to provide a tabbed property view in support of the
+ * OpenShift explorer
+ * 
+ * @author jeff.cantrill
+ *
+ */
+public class ExplorerViewTabbedPropertyAdapterFactory implements IAdapterFactory {
+	private static final String VIEW_ID = "org.jboss.tools.openshift.express.ui.explorer.expressConsoleView";
```
Comment (fbricon) [inline plugins/org.jboss.tools.openshift.common.ui/src/org/jboss/tools/openshift/internal/common/ui/explorer/ExplorerViewTabbedPropertyAdapterFactory.java]: there's OpenShiftUIUtils.OPENSHIFT_EXPLORER_VIEW_ID already

File: plugins/org.jboss.tools.openshift.common.ui/src/org/jboss/tools/openshift/internal/common/ui/utils/DateTimeUtils.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+			builder.append(parts[0]).append(" hrs.");
+		}
+		if(Integer.valueOf(parts[1]) > 0){
+			builder.append(" ").append(parts[1]).append(" min.");
+		}
+		if(Integer.valueOf(parts[2]) > 0){
+			builder.append(" ").append(parts[2]).append(" sec.");
+		}
+		if(builder.length() == 0) {
+			builder.append("Less than 0 sec.");
```
Comment (fbricon) [inline plugins/org.jboss.tools.openshift.common.ui/src/org/jboss/tools/openshift/internal/common/ui/utils/DateTimeUtils.java]: How about "just now."?

File: plugins/org.jboss.tools.openshift.common.ui/src/org/jboss/tools/openshift/internal/common/ui/utils/DateTimeUtils.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+	public static String formatSince(String value, TimeZone timezone) {
+		try {
+			Date date = parse(value);
+			DateFormat formatter = DateFormat.getDateTimeInstance(DateFormat.SHORT, DateFormat.FULL);
+			if(timezone != null) {
+				formatter.setTimeZone(timezone);
+			}
+			return formatter.format(date);
+		} catch (ParseException e) {
+			System.out.println(e);
```
Comment (fbricon) [inline plugins/org.jboss.tools.openshift.common.ui/src/org/jboss/tools/openshift/internal/common/ui/utils/DateTimeUtils.java]: yuck

File: plugins/org.jboss.tools.openshift.common.ui/src/org/jboss/tools/openshift/internal/common/ui/utils/DateTimeUtils.java [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+import java.text.DateFormat;
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.TimeZone;
+
+import org.apache.commons.lang.time.DurationFormatUtils;
+
+public class DateTimeUtils {
```
Comment (fbricon) [inline plugins/org.jboss.tools.openshift.common.ui/src/org/jboss/tools/openshift/internal/common/ui/utils/DateTimeUtils.java]: We might switch to http://www.ocpsoft.org/prettytime/ in the future, or not, if it's considered overkill.

File: plugins/org.jboss.tools.openshift.express.client/.project [hunk 6]
```diff
@@ -20,14 +20,8 @@
 			<arguments>
 			</arguments>
 		</buildCommand>
-		<buildCommand>
```
Comment (fbricon) [inline plugins/org.jboss.tools.openshift.express.client/.project]: not sure we want that change in

File: plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/explorer/OpenShiftExplorerLabelProvider.java [hunk 7]
```diff
@@ -109,14 +113,25 @@ public StyledString getStyledText(Object element) {
...
+			Deployment d = (Deployment) element;
+			return style(d.getService().getName(), formatRoute(d.getRoutes()));
+		}
 		return super.getStyledText(element);
 	}
 
+	private String formatRoute(Collection<IResourceUIModel> routes) {
+		if(routes.size() > 0) {
+			IRoute route = (IRoute)routes.iterator().next().getResource();
+			return String.format("%s%s", route.getHost(), route.getPath());
```
Comment (fbricon) [inline plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/explorer/OpenShiftExplorerLabelProvider.java]: are we sure the host and path will always be separated by a / ?

File: plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/handler/DeleteResourceHandler.java [hunk 8]
```diff
@@ -32,9 +33,15 @@
 
 	@Override
 	public Object execute(final ExecutionEvent event) throws ExecutionException {
-		ISelection selection = HandlerUtil.getCurrentSelection(event);
+		ISelection selection = HandlerUtil.getActivePart(event).getSite().getWorkbenchWindow().getSelectionService().getSelection();
```
Comment (fbricon) [inline plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/handler/DeleteResourceHandler.java]: why the change?

File: plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/handler/PodLogsHandler.java [hunk 9]
```diff
@@ -18,25 +18,27 @@
...
  * @author jeff.cantrill
  */
 public class PodLogsHandler extends AbstractOpenShiftCliHandler {
 	private static final String [] STATES = new String [] {"Running", "Succeeded", "Failed"};
-	private static final String BUILDNAME_ANNOTATION = "openshift.io/build.name";
 
 	@Override
 	protected void handleEvent(ExecutionEvent event){
-		ISelection selection = HandlerUtil.getCurrentSelection(event);
+		ISelection selection = HandlerUtil.getActivePart(event).getSite().getWorkbenchWindow().getSelectionService().getSelection();
```
Comment (fbricon) [inline plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/handler/PodLogsHandler.java]: why?

File: plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/models/DeploymentResourceMapper.java [hunk 10]
```diff
@@ -0,0 +1,461 @@
...
+		return buildConfigs;
+	}
+	
+	private <T extends IResource> List<T> load(String kind){
+		List<T> resources = conn.getResources(kind, project.getName());
+		resources.parallelStream().forEach(r->cache.put(getCacheKey(r), r));
+		return resources;
+	}
+	
+	private void mapBuildsToDeploymentConfigs(Collection<IBuild> builds) {
```
Comment (fbricon) [inline plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/models/DeploymentResourceMapper.java]: or builds.forEach(b -> mapBuildToDeploymentConfig(b)), if you hate brackets

File: plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/models/DeploymentResourceMapper.java [hunk 11]
```diff
@@ -0,0 +1,461 @@
...
+		for (IBuild build : builds) {
+			mapBuildToDeploymentConfig(build);
+		}
+	}
+	
+	private void mapBuildToDeploymentConfig(IBuild build) {
+		String imageRef = imageRef(build, project);
+		Collection<IDeploymentConfig> deploymentConfigs = imageRefToDeployConfigs.get(imageRef);
+		if(imageRefToDeployConfigs.containsKey(imageRef) && deploymentConfigs != null) {
```
Comment (fbricon) [inline plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/models/DeploymentResourceMapper.java]: imageRefToDeployConfigs.containsKey(imageRef)  is redundant here

**Comment by fbricon** [comment]
As a general  note, try to avoid using parallel streams unless it may help to solve some performance issues. http://stackoverflow.com/questions/20375176/should-i-always-use-a-parallel-stream-when-possible

File: plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/handler/DeleteResourceHandler.java (hunk 8 above)
Comment (jcantrill) [inline plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/handler/DeleteResourceHandler.java]: The former returns null when trying to select a row in the table.  This change should allow the handler to be used regardless if the event originates from the properties view or the explorer tree

File: plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/explorer/OpenShiftExplorerLabelProvider.java (hunk 7 above)
Comment (jcantrill) [inline plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/explorer/OpenShiftExplorerLabelProvider.java]: Replacing with route.getUrl()

**Comment by jcantrill** [comment]
@fbricon updated per your comments

File: plugins/org.jboss.tools.openshift.common.ui/src/org/jboss/tools/openshift/internal/common/ui/explorer/DefaultPropertySection.java (hunk 1 above)
Comment (jcantrill) [inline plugins/org.jboss.tools.openshift.common.ui/src/org/jboss/tools/openshift/internal/common/ui/explorer/DefaultPropertySection.java]: There is no 'isDispose' method

Comment (adietish) [inline plugins/org.jboss.tools.openshift.common.ui/src/org/jboss/tools/openshift/internal/common/ui/explorer/DefaultPropertySection.java]: no there isnt, but you could do 

```java
DisposeUtils.isDisposed(page.getControl())
```
?

File: plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/handler/DeleteResourceHandler.java (hunk 8 above)
Comment (adietish) [inline plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/handler/DeleteResourceHandler.java]: slightly shorter would be (no need to get the active part and site):

```java
HandlerUtil.getActiveWorkbenchWindow(event).getSelectionService().getSelection();
```

File: plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/handler/PodLogsHandler.java (hunk 9 above)
Comment (adietish) [inline plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/handler/PodLogsHandler.java]: see above

**Comment by jcantrill** [comment]
part of feature branch: feature_19178_deployment_view

**Comment by jcantrill** [comment]
Closing as merged into feature branch
